### PR TITLE
[AD-34] New keystore

### PR DESCRIPTION
**Pull request by epicallan**
**Description:**

**YT issue:** https://issues.serokell.io/issue/AD-34

**Checklist:**

- [x] Updated docs if necessary
  - [x] [README](README.md)
  - [x] [TUI usage guide](docs/usage-tui.md)
  - [x] [Configuration documentation](docs/configuration.md)
- [x] My code complies with the [style guide](docs/code-style.md)
- [x] Tested my changes if they modify code

### CODE REVIEW

**Comment by epicallan** [comment]
We are unfortunately still using `Bi` instances due to the fact that we cannot have  aeson JSON instances for a some data types from `cardano` since they are in non exposed modules.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                in UserSecret <$> pEncMap <*> pVersion <*> pure mempty
+
+{-------------------------------------------------------------------------------
+  Utilities
+-------------------------------------------------------------------------------}
+
+esksToMap :: [EncryptedSecretKey] -> Map (AddressHash PublicKey) EncryptedSecretKey
+esksToMap esks =
+    esks <&> (\encKey -> (eskToKey encKey, encKey))
+          & Map.fromList
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: How about

```haskell
esksToMap = Map.fromList . map (\encKey -> (eskToKey encKey, encKey))
```

?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+data UserSecret
+  = UserSecret
+  { _usWallets  :: Map (AddressHash PublicKey) EncryptedSecretKey
+  , _usVersion  :: !Version
+  , _usFilePath :: !FilePath
+  } deriving (Show)
+
+makeLenses ''UserSecret
+
+instance Default UserSecret where
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Since this instance is _always_ used like so
```haskell
def & usFilePath .~ fp
```
perhaps it would be better to provide a separate function, like `makeUserSecret :: FilePath -> UserSecret` or something like that?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+       , getUserSecretKeys
+       , usFilePath
+       , usWallets
+       , usVersion
+       ) where
+
+import Control.Lens.TH (makeLenses)
+import Data.Aeson
+  (FromJSON, ToJSON, Value(..), toJSON, parseJSON, withText, withObject, (.:), object, (.=))
+import Data.Default
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Please avoid wildcard imports.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 4]
```diff
@@ -0,0 +1,97 @@
+-- | system file operations related to key-store UserSecret
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Please use a capital letter at the beginning and a full stop at the end.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+module Ariadne.Wallet.Cardano.Kernel.Keystore.Util
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: This module is related to I/O operations on the new `UserSecret`. Perhaps a better name for this module would be something like `Persistence` or `IO`?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+-- | Creates empty temporary storage file for keyStore testing purposes
+getTempKeystorePath :: MonadIO m => m FilePath
+getTempKeystorePath = do
+    tempDir <- liftIO getTemporaryDirectory
+    let dir = tempDir </> "ariadne-test"
+    liftIO $ createDirectoryIfMissing False dir
+    liftIO $ createEmptyFile $ dir </> "keystore.json"
+
+createEmptyFile :: FilePath -> IO FilePath
+createEmptyFile fp = writeFile fp mempty $> fp
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Why would you return the parameter from the function unchanged?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+import Ariadne.Wallet.Cardano.Kernel.Keystore.Types
+  (UserSecret(..), StorageDecodingError(..), usWallets, usFilePath)
+
+
+lockFilePath :: FilePath -> FilePath
+lockFilePath = (<> ".lock")
+
+-- | Creates empty temporary storage file for keyStore testing purposes
+getTempKeystorePath :: MonadIO m => m FilePath
+getTempKeystorePath = do
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: It is perhaps better to wrap the whole function in `liftIO $ do`.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    mode <- liftIO $ PSX.fileMode <$> PSX.getFileStatus path
+    return $ PSX.intersectFileModes mode PSX.accessModes
+
+-- | Set mode disregard current mode of the file.
+setMode :: MonadIO m => PSX.FileMode -> FilePath -> m ()
+setMode mode path = liftIO $ PSX.setFileMode path mode
+
+-- | Set given mode if needed.
+ensureModeIs
+    :: (MonadIO m, CanLog m)
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/System/FileMode.hs]: Do I understand correctly that this function is missing the `IsPosix` constraint?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    -- hence no handler.
+    (tempPath, tempHandle) <-
+        openBinaryTempFile (takeDirectory path) (takeFileName path)
+
+    LBS.hPut tempHandle (encode us) `finally` hClose tempHandle
+    hClose tempHandle
+    renameFile tempPath path
+
+-- | Whether stored UserSecret has empty keys
+isEmptyUserSecret :: UserSecret -> Bool
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: This seems like the only function in this module that is not related to I/O. Furthermore, it is not used anywhere in this module. Perhaps it should be moved elsewhere?

File: ariadne/cardano/test/backend/Test/Spec/Keystore.hs [hunk 10]
```diff
@@ -4,12 +4,12 @@ module Test.Spec.Keystore (
...
   ) where
 
-import System.Directory (doesFileExist, getTemporaryDirectory, removeFile)
-import System.FilePath ((</>))
+import System.Directory (doesFileExist, removeFile)
 import System.IO.Error (IOError)
 
 import Test.Hspec (Spec, describe, it, shouldBe, shouldReturn)
 import Test.Hspec.QuickCheck (prop)
+import Test.Pos.Core.Arbitrary ()
```
Comment (Pastafarianist) [inline ariadne/cardano/test/backend/Test/Spec/Keystore.hs]: Why do you need this import?

File: ariadne/cardano/src/Ariadne/System/FileMode.hs [hunk 11]
```diff
@@ -0,0 +1,39 @@
+module Ariadne.System.FileMode
+       ( getAccessMode
+       , setMode
+       , ensureModeIs
+       , mode600
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/System/FileMode.hs]: Are you sure that all of these exports are needed? You may run `weeder` on the codebase to see which of them are never used outside of this module.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+mode600 :: PSX.FileMode
+mode600 = PSX.unionFileModes PSX.ownerReadMode PSX.ownerWriteMode
+
+-- | Return only the access part of the file mode (like owner:rw-, etc).
+getAccessMode :: MonadIO m => FilePath -> m PSX.FileMode
+getAccessMode path = do
+    mode <- liftIO $ PSX.fileMode <$> PSX.getFileStatus path
+    return $ PSX.intersectFileModes mode PSX.accessModes
+
+-- | Set mode disregard current mode of the file.
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/System/FileMode.hs]:
```suggestion
-- | Set mode regardless of the current mode of the file.
```

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+esksToMap :: [EncryptedSecretKey] -> Map (AddressHash PublicKey) EncryptedSecretKey
+esksToMap esks =
+    esks <&> (\encKey -> (eskToKey encKey, encKey))
+          & Map.fromList
+
+eskToKey :: EncryptedSecretKey -> AddressHash PublicKey
+eskToKey = addressHash . encToPublic
+
+getUserSecretKeys :: UserSecret -> [EncryptedSecretKey]
+getUserSecretKeys us = Map.elems (us ^. usWallets)
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Actually, `eskToKey` and `getUserSecretKeys` look a lot like functions that could be put in a `Util` module.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 14]
```diff
@@ -251,8 +206,8 @@ delete walletId (Keystore ks) = do
 -- | Returns all the 'EncryptedSecretKey' known to this 'Keystore'.
 toList :: Keystore -> IO [(WalletId, EncryptedSecretKey)]
 toList (Keystore ks) =
-    withMVar ks $ \(InternalStorage us) ->
-        pure $ map (first hashPubKeyToWalletId) $ toPairs $ us ^. usWallets
+    withMVar ks $ \us ->
+        pure $ map (first hashPubKeyToWalletId) $ Map.toList $ us ^. usWallets
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Why replace `toPairs` here?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs (hunk 6 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Damn mistake. 😞

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs (hunk 7 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: I agree. 👍

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+writeRaw path us = do
+    -- On POSIX platforms, openBinaryTempFile guarantees that the file
+    -- will be created with mode 600.
+    -- If openBinaryTempFile throws, we want to propagate this exception,
+    -- hence no handler.
+    (tempPath, tempHandle) <-
+        openBinaryTempFile (takeDirectory path) (takeFileName path)
+
+    LBS.hPut tempHandle (encode us) `finally` hClose tempHandle
+    hClose tempHandle
```
[the remark this answers is not in the document]
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Another Damn mistake. Shouldn't be there 😞.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs (hunk 9 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Agreed!

File: ariadne/cardano/test/backend/Test/Spec/Keystore.hs (hunk 10 above)
Comment (epicallan) [reply ariadne/cardano/test/backend/Test/Spec/Keystore.hs]: I was getting errors related to lack of  class instances from `Test.Pos.Core.Arbitrary () `.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs (hunk 13 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Good call

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs (hunk 14 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: I didn't realize then that they are essentially same thing. Will revert.

File: ariadne/cardano/src/Ariadne/Config/Wallet.hs [hunk 16]
```diff
@@ -23,7 +23,7 @@ import Ariadne.Util (postfixLFields)
 defaultWalletConfig :: FilePath -> WalletConfig
 defaultWalletConfig dataDir = WalletConfig
     { wcEntropySize = 16
-    , wcKeyfilePath = dataDir </> "secret-mainnet.key"
+    , wcKeyfilePath = dataDir </> "secret-mainnet.json"
```
[the remark this answers is not in the document]
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Config/Wallet.hs]: I think its better.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+import qualified Data.Text.Buildable
+import qualified Data.Map as Map (empty, fromList, elems)
+import qualified Serokell.Util.Base16 as S (encode, decode)
+
+newtype Version= Version { getVersion :: Int }
+  deriving (Show, Eq)
+
+-- | UserSecret data containing wallet keys that gets saved to file
+data UserSecret
```
[the remark this answers is not in the document]
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Makes sense. Will switch to `InternalStorage`

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs (hunk 13 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Turns we can't have `getUserSecretKeys` in a `Util` module due to `cyclic dependence issue` between the 2 modules. I added some other functions there though.

File: ariadne/cardano/src/Ariadne/System/Other.hs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+import System.Info (os)
+
+---------------------------------------------------------------------
+-- OS-dependent operations
+---------------------------------------------------------------------
+
+data PosixOperationsEnabled = PosixOperationsEnabled
+
+-- | Allowance for POSIX operations.
+type IsPosix = Given PosixOperationsEnabled
```
[the remark this answers is not in the document]
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/System/Other.hs]: Its if we care about not throwing IO exceptions in a `Non Posix OS`

File: ariadne/cardano/src/Ariadne/System/FileMode.hs (hunk 8 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/System/FileMode.hs]: I thought so to, but adding it there triggers `warning: [-Wredundant-constraints]`

File: ariadne/cardano/src/Ariadne/System/FileMode.hs (hunk 11 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/System/FileMode.hs]: removed the ones not used. The code snippet here hasn't updated to reflect that though.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs (hunk 8 above)
Comment (Pastafarianist) [reply ariadne/cardano/src/Ariadne/System/FileMode.hs]: Please see the discussion in #haskell in Slack. The whole approach with `IsPosix` is not going to work. I suggest that you get rid of all mentions of it.

File: ariadne/cardano/src/Ariadne/System/Other.hs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/cardano/src/Ariadne/System/Other.hs]: Please see the discussion in #haskell in Slack. The whole approach with IsPosix is not going to work. I suggest that you get rid of all mentions of it.

File: ariadne/cardano/src/Ariadne/Wallet/Backend/Restore.hs [hunk 19]
```diff
@@ -91,10 +91,7 @@ restoreFromKeyFile ::
     -> WalletRestoreType
     -> IO ()
 restoreFromKeyFile pwl face runCardanoMode mbWalletName path rType = do
-    keyFile <- BS.readFile path
-    us <- case decodeFull' keyFile of
-      Left e   -> throwM $ SecretsDecodingError path e
-      Right us -> pure us
+    us <- readInternalStorage path
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Backend/Restore.hs]: `istorage`?

File: ariadne/cardano/test/backend/Test/Spec/Keystore.hs (hunk 10 above)
Comment (Pastafarianist) [reply ariadne/cardano/test/backend/Test/Spec/Keystore.hs]: Strange that they were not required before.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs (hunk 8 above)
Comment (Martoon-00) [reply ariadne/cardano/src/Ariadne/System/FileMode.hs]: IIUC In that discussion in #haskell you tried to use `IsPosix` constraint to make sure that `whenPosix` _is not_ called when it _is not_ necessary. while its original purpose was to ensure `whenPosix` _is_ used when it _is_ necessary.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 20]
```diff
@@ -218,3 +210,7 @@ toList (Keystore ks) =
 
 walletIdToKey :: WalletId -> AddressHash PublicKey
 walletIdToKey (WalletIdHdSeq hdRootId) = view fromDb $ getHdRootId hdRootId
+
+-- | Whether stored InternalStorage has empty keys
+isEmptyInternalStorage :: InternalStorage -> Bool
+isEmptyInternalStorage istorage = null (istorage ^. isWallets)
```
[the remark this answers is not in the document]
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Configured it right. Shouldn't happen again.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    build (StorageDecodingError msg) =
+        "Failed to decode stored json: " <> bprint build msg
+
+instance ToJSON EncryptedSecretKey where
+  toJSON = String . S.encode . serialize'
+
+instance FromJSON EncryptedSecretKey where
+  parseJSON  = withText "EncryptedSecretKey" $ \encryptedText ->
+                let eitherDecode = either textFail pure . decodeFull'
+                in either textFail eitherDecode $ S.decode encryptedText
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Please don't indent code like this. Add a linebreak after `parseJSON =`, then add 4 spaces per indentation level. Also please make sure you use 4 spaces everywhere.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+  toJSON istorage =
+    object [ "keys"    .= getInternalStorageKeys istorage
+           , "version" .= (toText $ showVersion $ istorage ^. isVersion)
+           ]
+
+instance FromJSON InternalStorage where
+  parseJSON = withObject "InternalStorage" $ \v ->
+                let pEncMap  = esksToMap <$> v .: "keys"
+                    pVersion = either textFail pure . decodeVersion =<< v .: "version"
+                in InternalStorage <$> pEncMap <*> pVersion <*> pure mempty
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Same as above.

**Comment by Pastafarianist** [comment]
Also, please ensure that `stylish-haskell` is applied to all affected files before merging.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+import  Ariadne.Wallet.Cardano.Kernel.Keystore.Util (esksToMap)
+
+import qualified Data.Text.Buildable
+import qualified Data.Map as Map (elems, empty)
+import qualified Serokell.Util.Base16 as S (encode, decode)
+
+-- | InternalStorage data containing wallet keys that gets saved to file
+data InternalStorage
+  = InternalStorage
+  { _isWallets  :: Map (AddressHash PublicKey) EncryptedSecretKey
```
Comment (Pastafarianist) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]:
```suggestion
  { _isWallets  :: !(Map (AddressHash PublicKey) EncryptedSecretKey)
```

File: ariadne/cardano/test/backend/Test/Spec/Keystore.hs [hunk 24]
```diff
@@ -70,11 +66,11 @@ spec =
                     (fmap hash mbKey) `shouldBe` (Just (hash esk))
 
         prop "Inserts are persisted after releasing the keystore" $ monadicIO $ do
-            keystorePath <- getKeystorePath
+            keystorePath <- getTempKeystorePath
             (STB wid, STB esk) <- pick genKeypair
             run $ do
                 nukeKeystore keystorePath
-                Keystore.bracketKeystore KeepKeystoreIfEmpty keystorePath $ \keystore1 ->
+                Keystore.bracketKeystore KeepKeystoreIfEmpty keystorePath $ \keystore1 -> do
```
Comment (gromakovsky) [inline ariadne/cardano/test/backend/Test/Spec/Keystore.hs]: Redundant change.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+
+import Pos.Core (AddressHash, addressHash)
+import Pos.Crypto (EncryptedSecretKey(..), PublicKey, encToPublic)
+
+import qualified Data.Map as Map (fromList)
+
+esksToMap :: [EncryptedSecretKey] -> Map (AddressHash PublicKey) EncryptedSecretKey
+esksToMap = Map.fromList . map (\encKey -> (eskToKey encKey, encKey))
+
+eskToKey :: EncryptedSecretKey -> AddressHash PublicKey
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: Why is it called `eskToKey` if it returns hash?

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+    deriving (Show)
+
+instance Exception StorageDecodingError
+
+instance Buildable StorageDecodingError where
+    build (StorageDecodingError msg) =
+        "Failed to decode stored json: " <> bprint build msg
+
+instance ToJSON EncryptedSecretKey where
+  toJSON = String . S.encode . serialize'
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: Let's avoid orphan instances. This instance is not necessary.
You can define `eskToJSON = String . S.encode . serialize'` and apply `map eskToJSON` to the result of `getInternalStorageKeys`.
Similar thing can be done for `fromJSON`.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+       ) where
+
+import System.Info (os)
+import System.Wlog (logWarning, usingLoggerName, CanLog)
+
+import qualified System.Posix.Files as PSX
+import qualified System.Posix.Types as PSX (FileMode)
+
+-- | It's common practice to make keyfiles readable/writable for user only.
+-- For now keyfile is read-only
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/System/FileMode.hs]: > For now keyfile is read-only

What does it mean?

File: ariadne/cardano/src/Ariadne/System/FileMode.hs [hunk 28]
```diff
@@ -0,0 +1,47 @@
+module Ariadne.System.FileMode
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/System/FileMode.hs]: I think it's better to put this module into `ariadne-core`, because it doesn't depend on anything from Cardano.

File: ariadne/cardano/src/Ariadne/Wallet/Backend/Restore.hs [hunk 29]
```diff
@@ -91,10 +91,7 @@ restoreFromKeyFile ::
     -> WalletRestoreType
     -> IO ()
 restoreFromKeyFile pwl face runCardanoMode mbWalletName path rType = do
-    keyFile <- BS.readFile path
-    us <- case decodeFull' keyFile of
-      Left e   -> throwM $ SecretsDecodingError path e
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Backend/Restore.hs]: AFAIU, after this deletion `SecretsDecodingError` (defined in this file) is not used anymore, so it should be deleted.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 30]
```diff
@@ -117,23 +87,23 @@ bracketKeystore deletePolicy fp withKeystore =
...
 
 -- | Creates a legacy 'Keystore' using a 'bracket' pattern, where the
 -- initalisation and teardown of the resource are wrapped in 'bracket'.
 -- For a legacy 'Keystore' users do not get to specify a 'DeletePolicy', as
 -- the release of the keystore is left for the node and the legacy code
 -- themselves.
-bracketLegacyKeystore :: UserSecret -> (Keystore -> IO a) -> IO a
-bracketLegacyKeystore us withKeystore =
-    bracket (newLegacyKeystore us)
+bracketLegacyKeystore :: InternalStorage -> (Keystore -> IO a) -> IO a
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Let's get rid of legacy. That's the primary goal of the issue. Fortunately, this function seems to be unused, so it can be simply deleted.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 31]
```diff
@@ -263,3 +210,7 @@ toList (Keystore ks) =
 
 walletIdToKey :: WalletId -> AddressHash PublicKey
 walletIdToKey (WalletIdHdSeq hdRootId) = view fromDb $ getHdRootId hdRootId
+
+-- | Whether stored InternalStorage has empty keys
+isEmptyInternalStorage :: InternalStorage -> Bool
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Tbh, I do not understand why `getInternalStorageKeys` is in `Types` and this function is here. Both are simple pure functions which take `InternalStorage` and return something, so I think they should be in the same module :man_shrugging:

File: ariadne/cardano/test/backend/Test/Spec/Keystore.hs (hunk 24 above)
Comment (Pastafarianist) [reply ariadne/cardano/test/backend/Test/Spec/Keystore.hs]: I think it's OK as it makes the code more uniform.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs (hunk 25 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs]: I think I adopted from previous code, I have changed.

File: ariadne/cardano/src/Ariadne/System/FileMode.hs (hunk 27 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/System/FileMode.hs]: Corrected it to `-- For now keyfile is read and write only`

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs (hunk 31 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: The reason is to avoid cyclic dependencies. An excellent place to put `getInternalStorageKeys` would be `Ariadne.Wallet.Cardano.Kernel.Keystore.Util` but then it would lead to cyclic dependencies since it would require `InternalStorage` in `Ariadne.Wallet.Cardano.Kernel.Keystore.Types` . `isEmptyInternalStorage` would also have a similar issue

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 32]
```diff
@@ -153,47 +106,27 @@ bracketTestKeystore withKeystore =
...
 
 -- | Releases the underlying 'InternalStorage' and returns the updated
 -- 'InternalStorage' and the file on disk this storage lives in.
--- 'FilePath'.
 release :: InternalStorage -> IO FilePath
-release (InternalStorage us) = do
-    let fp = getUSPath us
-    writeUserSecret us
-    return fp
+release istorage = writeInternalStorage istorage $> istorage  ^. isFilePath
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: It seems to be the only place where `isFilePath` is actually used. And that can be easily avoided. `release` is only called from `releaseKeystore`. It's called from `keystoreComponent`, `bracketKeystore` and `bracketTestKeystore`. The first two can pass `FilePath` to `releaseKeystore` already. `bracketTestKeystore` needs to be changed a little bit so that `newTestKeystore` returns a pair.

I don't understand this design when `FilePath` is part of `Keystore`. I know it was like that before, but I'd like to get rid of that unless someone provides a good reason why it's good.

One reason why it's bad is that when we parse `JSON` we return `Keystore` with some static (empty) file path. That's already enough for me to dislike this design.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 33]
```diff
@@ -263,3 +193,7 @@ toList (Keystore ks) =
 
 walletIdToKey :: WalletId -> AddressHash PublicKey
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Btw, I have a similar question to the one I had about `eskToKey`. It returns a hash, not a key.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+import qualified Data.Map as Map (elems, empty)
+import qualified Data.Text.Buildable
+import qualified Data.Vector as V
+import qualified Serokell.Util.Base16 as S (decode, encode)
+
+-- | InternalStorage data containing wallet keys that gets saved to file
+data InternalStorage
+  = InternalStorage
+  { _isWallets  :: !(Map (AddressHash PublicKey) EncryptedSecretKey)
+  , _isVersion  :: !Version
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: We don't need this field in the data type. We don't use it anywhere. We should put `version` into serialized JSON value and check it during deserialization. But we shouldn't carry it further.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs [hunk 35]
```diff
@@ -0,0 +1,101 @@
...
+
+eskToJSON :: EncryptedSecretKey -> Value
+eskToJSON = String . S.encode . serialize'
+
+eskFromJson :: Value -> Parser EncryptedSecretKey
+eskFromJson (String encryptedText) =
+    let eitherDecode = either textFail pure . decodeFull'
+    in either textFail eitherDecode $ S.decode encryptedText
+
+eskFromJson _ = fail "Parsing EncryptedSecretKey Array error"
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Types.hs]: It's better to use `typeMismatch` from `aeson`.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs (hunk 32 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: @gromakovsky The reason why we have filePath as part of `keystore` is that it's passed as a configuration value from ` WalletConfig` in `wcKeyfilePath`

if we derive the filePath from within `Ariadne.Wallet.Cardano.Kernel.Keystore.Persistence` then we can get rid of filePath in `keystore`.

Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: @gromakovsky The reason is that it allows us to use keystore operations such as `insert` and `delete` without use of a `filePath` value in function arguments.

If we derived filePath from with in `Ariadne.Wallet.Cardano.Kernel.Keystore.Persistence` then it would be easy to get rid of filePath from with in `keystore`.

Comment (gromakovsky) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: Ok, I agree, I didn't notice that. However, instead of storing `FilePath` in `InternalStorage`, why not store it in `Keystore`? This way we won't have to set default value in JSON deserialization.

Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: That sounds good.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs (hunk 33 above)
Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: I changed the name accordingly.

File: ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs [hunk 36]
```diff
@@ -34,15 +35,24 @@ import Pos.Core (AddressHash, addressHash)
...
+  (InternalStorage(..), isWallets)
+import Ariadne.Wallet.Cardano.Kernel.Keystore.Util
+  (eskToPublicHash, walletIdToPublicHash)
 import Ariadne.Wallet.Cardano.Kernel.Types (WalletId(..))
 
-data Keystore = Keystore (MVar InternalStorage)
+data Storage
+  = Storage
+  { _sIstorage :: !InternalStorage
+  , _sFilePath :: !FilePath
```
Comment (gromakovsky) [inline ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: What's the purpose of putting `FilePath` into `MVar`? It's constant, why not make it a pure value?

Comment (epicallan) [reply ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs]: I didn't think through it deeply. I am going to change it such that I have this kind of constructor instead:

```
data Keystore
  = Keystore
  { _kstorage :: !(Mvar InternalStorage)
  , _sFilePath :: !FilePath
  }
```
